File: 프로그래머스/3/258705. 산 모양 타일링/산 모양 타일링.java
class Solution {

    private static final int MOD = 10_007;

    public int solution(int n, int[] tops) {
        int size = n * 3 + 2;

        int[] dp = new int[size];
        dp[1] = 1;
        dp[2] = 2;

        for (int i = 3; i < size; i++) {
            dp[i] = dp[i - 1];
            dp[i] += switch (i % 3) {
                case 0 -> (tops[(i / 3) - 1] == 1) ? dp[i - 2] : 0;
                case 1 -> dp[i - 3];
                case 2 -> (tops[(i - 1) / 3 - 1] == 1) ? dp[i - 2] : dp[i - 3];
                default -> 0;
            };
            dp[i] %= MOD;
        }

        return dp[size - 1];
    }
}

File: 프로그래머스/3/258705. 산 모양 타일링/산 모양 타일링.py
def solution(n, tops):
    size = n * 3 + 2
    
    dp = [0] * size
    dp[1] = 1
    dp[2] = 2

    for i in range(3, size):
        if i % 3 == 0:
            t = tops[(i // 3) - 1]
            dp[i] = dp[i - 1] + dp[i - 2] if t == 1 else dp[i - 1]
        elif i % 3 == 1:
            t = tops[(i // 3) - 1]
            dp[i] = dp[i - 1] + dp[i - 3] if t == 1 else dp[i - 2] + dp[i - 3]
        else:
            t = tops[((i - 1) // 3) - 1]
            dp[i] = dp[i - 1] + dp[i - 2] if t == 1 else dp[i - 1] + dp[i - 3]
        
        dp[i] %= 10007

    return dp[size - 1]